src/theory/sets/cardinality_extension.h

```diff
@@ -0,0 +1,343 @@
+   * Additionally, (A union B) is a parent of all three of the above sets
+   * if it exists as a term in the current context. As exceptions,
+   * if A op B = A, then A is not a parent of A ^ B and similarly for B.
+   * If A ^ B is empty, then it has no parents.
+   *
+   * We say the cardinality graph induced by the current set of equalities
+   * is an (irreflexive, acyclic) graph whose nodes are equivalence classes and
+   * which contains a (directed) edge r1 to r2 if there exists a term t2 in r2
+   * that has some parent t1 in r1.
+   *
+   * (2) d_oSetEqc, an ordered set of equivalence classes whose types are set.
+   * These equivalence classes have the property that if r1 is a descendant
+   * of r2 in the cardinality graph, then r1 must come before r2 in d_oSetEqc.
+   *
+   * This inference schema may make various inferences while building these
+   * two data structures if the current equality arrangement of sets is not
+   * as expected. For example, it will infer equalities between sets based on
+   * the emptiness and equalities of sets in adjacent children in the
+   * cardinality graph, to give some examples:
+   *   (A \ B = empty) => A = A ^ B
+   *   A^B = B => B \ A = empty
+   *   A union B = A ^ B => A \ B = empty AND B \ A = empty
+   * and so on.
+   *
+   * It will also recognize when a cycle occurs in the cardinality graph, in
+   * which case an equality chain between sets can be inferred. For an example,
+   * see checkCardCyclesRec below.
+   *
+   * This method is inspired by the checkCycles inference schema in the theory
+   * of strings.
+   */
+  void checkCardCycles();
+  /**
+   * Helper function for above. Called when wish to process equivalence class
+   * eqc.
+   *
+   * Argument curr contains the equivalence classes we are currently processing,
+   * which are descendants of eqc in the cardinality graph.
+   *
+   * Argument exp contains an explanation of why the chain of children curr
+   * are descedants of . For example, say we are in context with equivalence
+   * classes:
+   *   { A, B, C^D }, { D, B ^ C,  A ^ E }
+   * We may recursively call this method via the following steps:
+   *   eqc = D, curr = {}, exp = {}
+   *   eqc = A, curr = { D }, exp = { D = B^C }
+   *   eqc = A, curr = { D, A }, exp = { D = B^C, A = C^D }
+   * after which we discover a cycle in the cardinality graph. We infer
+   * that A must be equal to D, where exp is an explanation of the cycle.
+   */
+  void checkCardCyclesRec(Node eqc,
+                          std::vector<Node>& curr,
+                          std::vector<Node>& exp);
+  /** check normal forms
+   *
+   * This method attempts to assign "normal forms" to all set equivalence
+   * classes whose types have cardinality constraints. Normal forms are
+   * defined recursively.
+   *
+   * A "normal form" of an equivalence class [r] (where [r] denotes the
+   * equivalence class whose representative is r) is a set of representatives
```

Comment:
Isn't it an ordered set? the argument to the function is a vector.

Reply:
No, the order of the normal form doesn't matter.

`intro_sets` doesn't have to do with normal forms, see documentation at bottom of the function.